pipeline/src/resolver/mod.rs

```diff
@@ -0,0 +1,615 @@
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+actor_envelope! {
+    ResolverEnvelope,
+    ResolverActor,
+    ResolverRecorder,
+    // Subscribe to the stream_states table, we do not yet expose a subscription to stream_tips
+    SubscribeSince => SubscribeSinceMsg,
+    NewEventStates => NewEventStatesMsg,
+    StreamState => StreamStateMsg,
+}
+
+#[async_trait]
+impl Handler<SubscribeSinceMsg> for Resolver {
+    async fn handle(
+        &mut self,
+        message: SubscribeSinceMsg,
+    ) -> <SubscribeSinceMsg as Message>::Result {
+        let subscription = self.broadcast_tx.subscribe();
+        let ctx = self.ctx.clone();
+        rows_since(RowsSinceInput {
+            session_context: &ctx,
+            schema: schemas::stream_states(),
+            order_col: "stream_state_order",
+            projection: message.projection,
+            filters: message.filters.clone(),
+            limit: message.limit,
+            subscription: Box::pin(RecordBatchStreamAdapter::new(
+                schemas::stream_states(),
+                tokio_stream::wrappers::BroadcastStream::new(subscription)
+                    .map_err(|err| exec_datafusion_err!("{err}")),
+            )),
+            // Future Optimization can be to send the projection and limit into the events_since call.
+            since: stream_states_since(&ctx, message.filters).await?,
+        })
+    }
+}
+
+async fn stream_states_since(
+    ctx: &SessionContext,
+    filters: Option<Vec<Expr>>,
+) -> Result<SendableRecordBatchStream> {
+    let mut stream_states = ctx
+        .table(STREAM_STATES_TABLE)
+        .await?
+        .select(vec![wildcard()])?
+        // Do not return the partition columns
+        .drop_columns(&["stream_cid_partition"])?
+        .sort(vec![col("stream_state_order").sort(true, true)])?;
+
+    if let Some(filters) = filters {
+        for filter in filters {
+            stream_states = stream_states.filter(filter)?;
+        }
+    }
+    Ok(stream_states.execute_stream().await?)
+}
+
+/// Inform the resolver about new event states.
+#[derive(Debug)]
+pub struct NewEventStatesMsg {
+    events: RecordBatch,
+}
+impl Message for NewEventStatesMsg {
+    type Result = anyhow::Result<()>;
+}
+
+#[async_trait]
+impl Handler<NewEventStatesMsg> for Resolver {
+    async fn handle(
+        &mut self,
+        message: NewEventStatesMsg,
+    ) -> <NewEventStatesMsg as Message>::Result {
+        debug!(event_count = message.events.num_rows(), "new event states");
+        let batch = self.process_event_states_batch(message.events).await?;
+        if batch.num_rows() > 0 {
+            let _ = self.broadcast_tx.send(batch);
+        }
+        Ok(())
+    }
+}
+
+/// Request the state of a stream
+#[derive(Debug)]
+pub struct StreamStateMsg {
+    /// Id of the stream
+    pub id: StreamId,
+}
+impl Message for StreamStateMsg {
+    type Result = anyhow::Result<Option<StreamState>>;
+}
+
+/// State of a single stream
```

Comment:
I'm surprised we don't already have a type for this?  Don't we already have a stream_states table that must have a type for this?

Reply:
The pipeline naming is generally a bit confusing, and also, there are two instances of this struct, one in the Aggregator (`aggregator.rs`) and another in the Resolver (`resolver.rs`).

In the Aggregator (which applies commit patches), it's used with the `event_states` table, while in the Resolver, it's used with the `stream_states` table.

I could rename it to `EventState` in the Aggregator. It isn't _really_ the _stream_ state at that stage of the pipeline because the true stream state can only be determined after conflict resolution in the Resolver.

WDYT?

Reply:
definitely think they should have different names if they have different meanings.  Maybe this one is `ResolvedStreamState` or something like that?